service/worker/indexer/esProcessor.go

```diff
@@ -0,0 +1,265 @@
+		mapToKafkaMsg collection.ConcurrentTxMap // used to map ES request to kafka message. key: esDocID+visType
+		config        *Config
+		logger        bark.Logger
+		metricsClient metrics.Client
+	}
+)
+
+var _ ESProcessor = (*esProcessorImpl)(nil)
+
+const (
+	// retry configs for es bulk processor
+	esProcessorInitialRetryInterval = 200 * time.Millisecond
+	esProcessorMaxRetryInterval     = 20 * time.Second
+)
+
+// NewESProcessorAndStart create new ESProcessor and start
+func NewESProcessorAndStart(config *Config, client *elastic.Client, processorName string,
+	logger bark.Logger, metricsClient metrics.Client) (ESProcessor, error) {
+	p := &esProcessorImpl{
+		config: config,
+		logger: logger.WithFields(bark.Fields{
+			logging.TagWorkflowComponent: logging.TagValueIndexerESProcessorComponent,
+		}),
+		metricsClient: metricsClient,
+	}
+
+	processor, err := client.BulkProcessor().
+		Name(processorName).
+		Workers(config.ESProcessorNumOfWorkers()).
+		BulkActions(config.ESProcessorBulkActions()).
+		BulkSize(config.ESProcessorBulkSize()).
+		FlushInterval(config.ESProcessorFlushInterval()).
+		Backoff(elastic.NewExponentialBackoff(esProcessorInitialRetryInterval, esProcessorMaxRetryInterval)).
+		After(p.bulkAfterAction).
+		Do(context.Background())
+	if err != nil {
+		return nil, err
+	}
+
+	p.processor = processor
+	p.mapToKafkaMsg = collection.NewShardedConcurrentTxMap(1024, p.hashFn)
+	return p, nil
+}
+
+func (p *esProcessorImpl) Stop() {
+	p.processor.Stop()
+	p.mapToKafkaMsg = nil
+}
+
+// Add an ES request, and an map item for kafka message
+func (p *esProcessorImpl) Add(request elastic.BulkableRequest, key string, kafkaMsg messaging.Message) {
+	if p.mapToKafkaMsg.Contains(key) {
+		kafkaMsg.Ack() // duplicate message
+		return
+	}
+	p.mapToKafkaMsg.Put(key, kafkaMsg)
+	p.processor.Add(request)
+}
+
+// bulkAfterAction is triggered after bulk processor commit
+func (p *esProcessorImpl) bulkAfterAction(id int64, requests []elastic.BulkableRequest, response *elastic.BulkResponse, err error) {
+	if err != nil {
+		// This happens after configured retry, which means something bad happens on cluster or index
+		// When cluster back to live, processor will re-commit those failure requests
+		p.logger.WithFields(bark.Fields{
+			logging.TagErr: err,
+		}).Error("Error commit bulk request.")
+
+		for _, request := range requests {
+			p.logger.WithFields(bark.Fields{
+				logging.TagESRequest: request.String(),
+			}).Error("ES request failed.")
+
+			p.metricsClient.IncCounter(metrics.ESProcessorScope, metrics.ESProcessorFailures)
+		}
+		return
+	}
+
+	responseItems := response.Items
+	for i := 0; i < len(requests); i++ {
+		visType := p.getVisibilityType(requests[i])
+		if visType == "" {
+			continue
+		}
+		responseItem := responseItems[i]
+		for _, resp := range responseItem {
+			key := resp.Id + visType
+			switch {
+			case isResponseSuccess(resp.Status):
+				p.ackKafkaMsg(key)
+			case !isResponseRetriable(resp.Status):
```

Comment:
I'm not sure how the action is getting retried.

Reply:
Every requests added to ElasticBulkProcessor will be removed from queue after successfully committed, or failed with non-retriable status code (configurable). All the other requests not removed will stay in queue and retry forever.